From 830984a2dc20e318baeca66b20289d51fe3d0c5b Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 1 Oct 2024 09:18:13 +0200
Subject: [PATCH 1267/1695] core: dt: add fdt_reg_base_ncells()

Add fdt_reg_base_ncells() helper function to read DT reg property
when the number of cells to read is known. This helper function
speeds reading of DT "reg" properties since there is no need to
lookup parent node to get its #address-cells property value.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d871ad2dd273595a143a1422e9a911f752ff445a core: dt: add fdt_reg_base_ncells()]
Change-Id: Idd3a7b681c475b3bb255b5c49afc311f99b58bfe
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/408093
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/include/kernel/dt.h | 15 +++++++++++++++
 core/kernel/dt.c         | 17 +++++++++++++++++
 2 files changed, 32 insertions(+)

diff --git a/core/include/kernel/dt.h b/core/include/kernel/dt.h
index f163a379e..caedc5cbc 100644
--- a/core/include/kernel/dt.h
+++ b/core/include/kernel/dt.h
@@ -173,6 +173,14 @@ paddr_t fdt_read_paddr(const uint32_t *cell, int n);
  */
 paddr_t fdt_reg_base_address(const void *fdt, int offs);
 
+/*
+ * Return the base address for the "reg" property of the specified node
+ * expecting "reg" is made of @ncells FDT cells.
+ * Allowed values for @ncells are 1 and 2.
+ * Return (paddr_t)-1 in case of error.
+ */
+paddr_t fdt_reg_base_ncells(const void *fdt, int offs, size_t ncells);
+
 /*
  * Return the reg size for the reg property of the specified node or -1 in case
  * of error
@@ -335,6 +343,13 @@ static inline paddr_t fdt_reg_base_address(const void *fdt __unused,
 	return (paddr_t)-1;
 }
 
+static inline paddr_t fdt_reg_base_ncells(const void *fdt __unused,
+					  int offs __unused,
+					  size_t ncells __unused)
+{
+	return (paddr_t)-1;
+}
+
 static inline size_t fdt_reg_size(const void *fdt __unused,
 				  int offs __unused)
 {
diff --git a/core/kernel/dt.c b/core/kernel/dt.c
index 8d7344f59..9da607007 100644
--- a/core/kernel/dt.c
+++ b/core/kernel/dt.c
@@ -173,6 +173,23 @@ bad:
 
 }
 
+paddr_t fdt_reg_base_ncells(const void *fdt, int offs, size_t ncells)
+{
+	paddr_t reg = 0;
+	int len = 0;
+
+	/* fdt_read_paddr() traps most unexpected len values */
+	reg = fdt_read_paddr(fdt_getprop(fdt, offs, "reg", &len),
+			     len / sizeof(uint32_t));
+	if ((size_t)len != ncells * sizeof(uint32_t)) {
+		EMSG("Wrong reg cells number on node %s",
+		     fdt_get_name(fdt, offs, NULL));
+		return DT_INFO_INVALID_REG;
+	}
+
+	return reg;
+}
+
 paddr_t fdt_reg_base_address(const void *fdt, int offs)
 {
 	const void *reg;
-- 
2.47.2

